### PR TITLE
[Advanced Compiler]Add TE.scaled_softmax forward&backward

### PR DESCRIPTION
**Pull request by AdvancedCompiler**
### PR Category

Operator
### Type of Change

New Feature
### Description

Add transformer_engine_torch.scaled_softmax_forward&transformer_engine_torch.scaled_softmax_backward
### Progress

- [ ] Change is properly reviewed (1 reviewer required, 2 recommended).
- [ ] Change is responded to an issue.
- [ ] Change is fully covered by a UT.

### Performance

7c8c27d18ed7ef1e81ee70d107d38f81

### CODE REVIEW

File: src/flag_gems/ops/scaled_softmax.py [hunk 1]
```diff
@@ -0,0 +1,279 @@
...
+    def grid(meta):
+        BLOCK_Q = meta["BLOCK_Q"]
+        query_seq_tile_len = triton.cdiv(query_seq_len, BLOCK_Q)
+        return (query_seq_tile_len, attn_heads, batch_size)
+
+    grad_input = torch.empty_like(grad_output)
+
+    stride_b = softmax_results.stride(0)
+    stride_h = softmax_results.stride(1)
+    stride_q = softmax_results.stride(2)
```
Comment (huangyiqun) [inline src/flag_gems/ops/scaled_softmax.py:265]: Should the stride here be the same as the stride from grad_output?

File: benchmark/test_reduction_perf.py [hunk 2]
```diff
@@ -397,3 +397,77 @@ def test_quantile_benchmark(op_name, torch_op, input_fn, dtypes):
...
+    try:
+        from transformer_engine.common import load_framework_extension
+
+        load_framework_extension("torch")
+        import transformer_engine_torch as tex  # type: ignore
+    except ImportError:
+        pytest.skip("TransformerEngine is not available, skipping performance test")
+
+    def scaled_softmax_backward_input_fn(shape, dtype, device):
+        P = generate_tensor_input(shape, dtype, device)
```
Comment (huangyiqun) [inline benchmark/test_reduction_perf.py]: Using randomly generated input here may present problems; the output of the actual softmax function has specific mathematical properties.

Comment (huangyiqun) [reply benchmark/test_reduction_perf.py]: Maybe can use PyTorch to generate real softmax output.

File: src/flag_gems/ops/scaled_softmax.py (hunk 1 above)
Comment (songdejun) [reply src/flag_gems/ops/scaled_softmax.py:265]: The grad_output is `dL / dP` while L is the loss function.
The softmax_results is P with expression:
`P = softmax(S / √d)`
There's no shape change between S and P. 
And the grad_input is `dL / dS` which has the same shape with S, i.e. `[batch, head, seq, head_dim]`.
Thus:
`shape(grad_output) = shape(P) = shape(S) = shape(grad_input)`

File: benchmark/test_reduction_perf.py (hunk 2 above)
Comment (songdejun) [reply benchmark/test_reduction_perf.py]: The input data has been generated by torch.softmax.

**Review by huangyiqun — COMMENTED** [review]
The CI test on Hopper has been updated. Please merge the latest master to trigger the correct CI run.

**Review by huangyiqun — APPROVED** [review]

File: benchmark/test_reduction_perf.py [hunk 3]
```diff
@@ -476,3 +476,78 @@ def test_quantile_benchmark(op_name, torch_op, input_fn, dtypes):
...
+            yield from self.input_fn(shape, cur_dtype, self.device)
+
+
+@pytest.mark.scaled_softmax
+def test_perf_scaled_softmax_forward():
+    try:
+        from transformer_engine.common import _load_library
+
+        _load_library()
+        import transformer_engine_torch as tex  # type: ignore
```
Comment (awayzjj) [inline benchmark/test_reduction_perf.py:510]: Hello, I was wondering why we don’t use
`from transformer_engine.pytorch import cpp_extensions as tex`,
as done in just like https://github.com/flagos-ai/FlagGems/blob/master/tests/test_unary_pointwise_ops.py#L8.
Is there a specific reason for using a different import pattern here?
